module/zfs/zio.c

```diff
@@ -143,12 +143,21 @@ zio_init(void)
 		 */
 		if (arc_watch && !IS_P2ALIGNED(size, PAGESIZE))
 			continue;
-#endif
+		/*
+		 * Here's the problem - on 4K native devices in userland on
+		 * Linux using O_DIRECT, buffers must be 4K aligned or I/O
+		 * will fail with EINVAL, causing zdb (and others) to coredump.
+		 * Since userland probably doesn't need optimized buffer caches,
+		 * we just force 4K alignment on everything.
+		 */
+		align = PAGESIZE;
```

Comment:
> Maybe this should be 4096 explicitly, instead of PAGESIZE. 

That's not a bad idea, assuming PAGESIZE in 4096 is a bit x86_64 centric.

There is a similar restriction for the kernel slabs but the per-disk ashift is used for the alignment.  The biovec which is constructed for the IO maps the appropriate portions of the pages which make up the slab buffer in to the bios.